src/game/common/bezfwditerator.cpp

```diff
@@ -0,0 +1,106 @@
+/**
+ * @file
+ *
+ * @author xezon
+ *
+ * @brief Bezier Forward Iterator class.
+ *
+ * @copyright Thyme is free software: you can redistribute it and/or
+ *            modify it under the terms of the GNU General Public License
+ *            as published by the Free Software Foundation, either version
+ *            2 of the License, or (at your option) any later version.
+ *            A full copy of the GNU General Public License can be found in
+ *            LICENSE
+ */
+#include "bezfwditerator.h"
+
+BezFwdIterator::BezFwdIterator()
+{
+    m_index = 0;
+    m_count = 0;
+    m_point.Zero();
+    m_add0.Zero();
+    m_add1.Zero();
+    m_add2.Zero();
+}
+
+BezFwdIterator::BezFwdIterator(int count, const BezierSegment *segment)
+{
+    m_point.Zero();
+    m_add0.Zero();
+    m_add1.Zero();
+    m_add2.Zero();
+    m_index = 0; // Thyme specific.
+    m_count = count;
+    m_segment = *segment;
+}
+
+void BezFwdIterator::Start()
+{
+    m_index = 0;
+
+    if (m_count > 1) {
+        const float distance = 1.0f / (m_count - 1);
+        const float sqr_distance = distance * distance;
+        const float cub_distance = distance * distance * distance;
+
+        D3DXVECTOR4 xvec(m_segment.m_points[0].x, m_segment.m_points[1].x, m_segment.m_points[2].x, m_segment.m_points[3].x);
+        D3DXVECTOR4 yvec(m_segment.m_points[0].y, m_segment.m_points[1].y, m_segment.m_points[2].y, m_segment.m_points[3].y);
+        D3DXVECTOR4 zvec(m_segment.m_points[0].z, m_segment.m_points[1].z, m_segment.m_points[2].z, m_segment.m_points[3].z);
+
+        D3DXVECTOR4 mtx[3];
+        D3DXVec4Transform(&mtx[0], &xvec, &BezierSegment::s_bezBasisMatrix);
+        D3DXVec4Transform(&mtx[1], &yvec, &BezierSegment::s_bezBasisMatrix);
+        D3DXVec4Transform(&mtx[2], &zvec, &BezierSegment::s_bezBasisMatrix);
+
+        m_point = m_segment.m_points[0];
+
+        float *f1;
+        float *f2;
+        float *f3;
+        int idx = 3;
+
+        while (idx-- != 0) {
+            const float x = mtx[idx].x;
+            const float y = mtx[idx].y;
+            switch (idx) {
+                case 0:
+                    f1 = &m_add0.x;
+                    f2 = &m_add1.x;
+                    f3 = &m_add2.x;
+                    break;
+                case 1:
+                    f1 = &m_add0.y;
+                    f2 = &m_add1.y;
+                    f3 = &m_add2.y;
+                    break;
+                case 2:
+                    f1 = &m_add0.z;
+                    f2 = &m_add1.z;
+                    f3 = &m_add2.z;
+                    break;
+            }
+            *f1 = x * cub_distance + y * sqr_distance + mtx[idx].z * distance;
+            *f2 = 6.0f * x * cub_distance + 2.0f * y * sqr_distance;
+            *f3 = 6.0f * x * cub_distance;
```

Comment:
Mathematically this looks suspect. I would have expected f2 to be 3*cube and f3 6*cube. This is because it looks like this is differentiation of the line above. Having not much knowledge of these curves though it may be correct.

Reply:
Disregard above its doing something else.